From oliver@neukum.org  Thu Jul 31 10:18:34 2008
From: Oliver Neukum <oliver@neukum.org>
Date: Fri, 25 Jul 2008 16:19:39 +0200
Subject: USB: fix deadlock in cdc-acm
To: Greg KH <greg@kroah.com>
Cc: linux-usb@vger.kernel.org
Message-ID: <200807251619.39813.oliver@neukum.org>
Content-Disposition: inline


this fixes a deadlock in cdc-acm caused by disconnect() releasing the
second interface.

Signed-off-by: Oliver Neukum <oneukum@suse.de>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 drivers/usb/class/cdc-acm.c |    8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

--- a/drivers/usb/class/cdc-acm.c
+++ b/drivers/usb/class/cdc-acm.c
@@ -1221,11 +1221,11 @@ static void acm_disconnect(struct usb_in
 	struct acm *acm = usb_get_intfdata(intf);
 	struct usb_device *usb_dev = interface_to_usbdev(intf);
 
-	mutex_lock(&open_mutex);
-	if (!acm || !acm->dev) {
-		mutex_unlock(&open_mutex);
+
+	if (!acm || !acm->dev)
 		return;
-	}
+
+	mutex_lock(&open_mutex);
 	if (acm->country_codes){
 		device_remove_file(&acm->control->dev,
 				&dev_attr_wCountryCodes);
